### PR TITLE
Allow pre-compiled binaries for ruby 3.1.0

### LINKED ISSUE #9364: Add Ruby 3.1 compability
[Ruby 3.1.0 was released](https://www.ruby-lang.org/en/news/2021/12/25/ruby-3-1-0-released/) on Christmas. 

Today I started updating our apps and noticed that `google-protobuf` is not compatible with Ruby 3.1 yet.

```
google-protobuf-3.19.1-x86_64-darwin requires ruby version < 3.1.dev, >= 2.3, which is incompatible with 
the current version, ruby 3.1.0p0
```

I understand that it might take some time to get this fixes – especially during the end of year break. I created this issue to keep track of this and me notified when a updated version was released.

Cheers and Happy New Year.

### PR DESCRIPTION
**Pull request by Maaarcocr**
## Why?

Fixes #9364

## What are the changes? 

I added 3.1.1 as a RUBY_CC_VERSION. I had to upgrade rake-compiler-dock to 1.2.0 as this is the only version of this gem that supports compilation for ruby 3.1+

Furthermore, I had to add a bit of logic (and a new platform this compiles to) as since RubyInstaller 3.1 x64 Windows is not `x64-mingw32` but rather `x64-mingw-ucrt` as explained [here](https://rubyinstaller.org/2021/12/31/rubyinstaller-3.1.0-1-released.html) 

Also `x64-mingw-ucrt` only supports 3.1+ whereas `x64-mingw32` only supports anything up to 3.0 so a little if statement is needed.

### CODE REVIEW

**Comment by Maaarcocr** [comment]
@elharo I saw you added the `ruby` label, could you please also add `release notes: yes` as a label? (I think it makes sense to add the PR description to the next release notes, as it would let people now that this package now ships with 3.1 compatible compiled shims). 

I cannot add any labels as I do not have the privileges to do so :/

File: ruby/Rakefile
Comment (rochlefebvre) [inline ruby/Rakefile:128]: I notice that the other versions are minor versions with a .0 as the patch, e.g. 3.0.0, 2.7.0... Your Ruby 3.1 version string is written as 3.1.1 instead of 3.1.0.

Is that intentional?

Comment (Maaarcocr) [reply ruby/Rakefile:128]: I went for the latest version of ruby. I'm not sure if it would make any difference, I could use 3.1.0 if it also would support 3.1.1 (not sure about this)

File: ruby/google-protobuf.gemspec
Comment (rochlefebvre) [inline ruby/google-protobuf.gemspec:26]: rake-compiler does not have a 1.2.0 release yet: https://rubygems.org/gems/rake-compiler.  Did you mean to move up to the latest release, 1.1.9?

File: ruby/Rakefile
Comment (rochlefebvre) [reply ruby/Rakefile:128]: I think it's worth a shot, if only for consistency.

File: ruby/google-protobuf.gemspec
Comment (Maaarcocr) [reply ruby/google-protobuf.gemspec]: I think I may have just mistakenly changed this. Fixed it back to what it was, as it also should support the latest (not that it's needed I think)

**Comment by Maaarcocr** [comment]
I did not noticed that more work would be needed to setup CI for both testing on ruby 31 and releasing. I've added the changes to fix this in my latest PR.

**Comment by rochlefebvre** [comment]
Looking at the logs for [MacOSRuby Release](https://source.cloud.google.com/results/invocations/48c787ae-e162-44fa-86dd-2e94a00165bf/targets/protobuf%2Fgithub%2Fmaster%2Fmacos%2Fruby_release%2Fpresubmit/log) I see this bit, toward the end:
```
installing bundled gems:            /Users/kbuilder/.rake-compiler/ruby/x86_64-darwin/ruby-3.1.0/lib/ruby/gems/3.1.0
                                    minitest 5.15.0
                                    power_assert 2.0.1
                                    rake 13.0.6
                                    test-unit 3.5.3
                                    rexml 3.2.5
                                    rss 0.2.9
                                    net-ftp 0.1.3
                                    net-imap 0.2.2
                                    net-pop 0.1.1
                                    net-smtp 0.3.1
                                    matrix 0.4.2
                                    prime 0.1.2
                                    rbs 2.0.0
Building native extensions. This could take a while...
/Users/kbuilder/.rake-compiler/sources/ruby-3.1.0/lib/rubygems/ext/builder.rb:95:in `run': ERROR: Failed to build gem native extension. (Gem::Ext::BuildError)

    current directory: /Users/kbuilder/.rake-compiler/ruby/x86_64-darwin/ruby-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
/Users/kbuilder/.rvm/rubies/ruby-2.7.0/bin/ruby --disable\\=gems -I/Users/kbuilder/.rake-compiler/builds/x86_64-darwin/ruby-3.1.0 -I /Users/kbuilder/.rake-compiler/sources/ruby-3.1.0/lib -r ./siteconf20220302-86972-osxhy4.rb extconf.rb
/Users/kbuilder/.rake-compiler/builds/x86_64-darwin/ruby-3.1.0/rbconfig.rb:13:in `<module:RbConfig>': ruby lib version (3.1.0) doesn't match executable version (2.7.0) (RuntimeError)
	from /Users/kbuilder/.rake-compiler/builds/x86_64-darwin/ruby-3.1.0/rbconfig.rb:11:in `<top (required)>'
	from /Users/kbuilder/.rake-compiler/ruby/x86_64-darwin/ruby-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension/siteconf20220302-86972-osxhy4.rb:1:in `require'
	from /Users/kbuilder/.rake-compiler/ruby/x86_64-darwin/ruby-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension/siteconf20220302-86972-osxhy4.rb:1:in `<top (required)>'
	from extconf.rb:in `require'

extconf failed, exit code 1
```
While building native extensions for gems bundled in Ruby 3.1, it complains that we're running a 2.7 interpreter with a 3.1 library.

I see that [ruby_build_environment.sh](https://github.com/protocolbuffers/protobuf/blob/master/kokoro/release/ruby/macos/ruby/ruby_build_environment.sh) defaults to Ruby 2.7 in a few spots.  I'm unsure if that is related somehow?

**Comment by Maaarcocr** [comment]
> While building native extensions for gems bundled in Ruby 3.1, it complains that we're running a 2.7 interpreter with a 3.1 library.
> 
> I see that [ruby_build_environment.sh](https://github.com/protocolbuffers/protobuf/blob/master/kokoro/release/ruby/macos/ruby/ruby_build_environment.sh) defaults to Ruby 2.7 in a few spots. I'm unsure if that is related somehow?

yeah I'm trying to understand how we build 3.0.0 and what's missing to get 3.1.0 working

**Comment by Maaarcocr** [comment]
I think that maybe I was overthinking the `ucrt` changes. I simplified the code to get rid of any logic specific to that and also made sure to run the 3.1 build for macos under ruby 3.1

**Comment by Maaarcocr** [comment]
It seems like rvm has not published a stable version that has support for 3.1.0

I'll look into either installing 3.1.0 from the ftp or we could wait for them to release a version with 3.1.0 support.

**Comment by Maaarcocr** [comment]
> It might be OK to temporarily use the rvm head and later go back to stable once 3.1.0 is available there.

Done!

**Comment by Maaarcocr** [comment]
I'm quite confused as to where rvm gets installed for the macos release machine. I can't seem to make it install the latest version

**Comment by Maaarcocr** [comment]
Ok so it looks like cross-compiling for arm64 is currently not working because it tries to assemble some arm64 assembly but it fails at doing so.

**Comment by Maaarcocr** [comment]
I figured out why compilation of ruby 3.1.0 was failing in CI. The configure script for ruby has changes such that it now replaces `aarch64-darwin` with `armd64-darwin` (https://github.com/ruby/ruby/blob/1abd068445aa74ed1d9483e319b5cf1d87b267f0/configure.ac#L118) which gives the actually correct coroutine implementation for arm64, rather than defaulting to the `ucontext` implementation. This change results in CI failing to compile assembly that is target to arm64 while compiling on an x86-64 machine. 

I decided that for 3.1.0 we can use `ucontext` for both arm64 and x86-64, so that CI passes (also it's more consistent). If anyone has any idea on how to this without having a new patched file, let me know.

**Comment by Maaarcocr** [comment]
In the meantime, I was trying to figure out how to fix the linux release build, as maven was failing to run. I made a pr to rake-compiler-dock, as their current jruby docker image contains a version of maven that does not work with the version of jdk they use. https://github.com/rake-compiler/rake-compiler-dock/pull/64

Hopefully I can get this merged soon and then I'll probably have to update this PR to use the new version.

**Comment by rochlefebvre** [comment]
🎉 Great work @Maaarcocr on getting the Mac Ruby Release check passing! 
image
Do we know why there is no _MacOS Ruby 3.1_ check listed here? Is it running somewhere but not registering with GH? Or do we need to add Ruby 3.1 to a testing matrix?

There is no _Linux Ruby 3.1_ either.

**Comment by Maaarcocr** [comment]
> Do we know why there is no _MacOS Ruby 3.1_ check listed here? Is it running somewhere but not registering with GH? Or do we need to add Ruby 3.1 to a testing matrix?
> 
> There is no _Linux Ruby 3.1_ either.

I'm quite confused as to why they do not appear, as I added the necessary files (I think). It's my first PR to anything that uses kokoro, so I'm not sure. If anyone with more experience with this particular CI setup could confirm that the ruby 3.1 tests will show up at once merged, that'd be great! (Or if I'm missing something)

**Comment by acozzette** [comment]
Sorry for the confusion on this, there is a Google-internal config we need to add before the 3.1 test runs are fully enabled. I can take care of this after the PR is merged.

**Comment by acozzette** [comment]
@Maaarcocr I rebuilt the Docker image, so that should fix most or all of the failing test runs. The change looks good to me overall but I want to look more closely at the changes to ruby_build_environment.sh on Monday. Thank you for all your work on this!

**Review by acozzette — APPROVED** [review]
Looks good to me but I guess we will still need to wait for your rake-compiler-dock change to be released.

**Comment by haberman** [comment]
> I tried this, but the issue is that the docker image that is pulled still does not show the changes I made (I guess someone has to push a new image version).

Is this `kokoro/linux/dockerfile/test/ruby/Dockerfile`?

I pulled your branch and tried running `kokoro/linux/dockerfile/push_testing_images.sh`, which usually updates our Docker images, but it didn't affect this one.

**Comment by haberman** [comment]
When I look at the latest test runs, everything seems to have passed except for "Linux Ruby Release ", which failed in a JRuby step (attn: @JasonLunn ).

```
[ERROR] Error executing Maven.
[ERROR] java.lang.IllegalStateException: Unable to load cache item
[ERROR] Caused by: Unable to load cache item
[ERROR] Caused by: Could not initialize class com.google.inject.internal.cglib.core.$MethodWrapper
rake aborted!
Command failed with status (1): [docker run -v /tmpfs/src/github/protobuf:/tmpfs/src/github/protobuf -e UID\=1000 -e GID\=1000 -e USER\=kbuilder -e GROUP\=kbuilder -e GEM_PRIVATE_KEY_PASSPHRASE -e ftp_proxy -e http_proxy -e https_proxy -e RCD_HOST_RUBY_PLATFORM\=x86_64-linux -e RCD_HOST_RUBY_VERSION\=2.4.1 -e RCD_IMAGE\=larskanis/rake-compiler-dock-jruby:1.2.0 -w /tmpfs/src/github/protobuf --rm -i larskanis/rake-compiler-dock-jruby:1.2.0 runas sigfw bash -c \ \ \ \ \ \ sudo\ apt-get\ install\ maven\ -y\ \&\&\ \ \ \ \ \ \ cd\ java\ \&\&\ mvn\ install\ -Dmaven.test.skip\=true\ \&\&\ cd\ ../ruby\ \&\&\ \ \ \ \ \ \ bundle\ \&\&\ \ \ \ \ \ \ IN_DOCKER\=true\ rake\ compile\ gem'
']
/usr/local/rvm/gems/ruby-2.4.1/bundler/gems/rake-compiler-dock-cd2afa24deca/lib/rake_compiler_dock/starter.rb:97:in `block in exec'
/usr/local/rvm/gems/ruby-2.4.1/bundler/gems/rake-compiler-dock-cd2afa24deca/lib/rake_compiler_dock/starter.rb:47:in `each'
/usr/local/rvm/gems/ruby-2.4.1/bundler/gems/rake-compiler-dock-cd2afa24deca/lib/rake_compiler_dock/starter.rb:47:in `exec'
/usr/local/rvm/gems/ruby-2.4.1/bundler/gems/rake-compiler-dock-cd2afa24deca/lib/rake_compiler_dock/starter.rb:14:in `sh'
/usr/local/rvm/gems/ruby-2.4.1/bundler/gems/rake-compiler-dock-cd2afa24deca/lib/rake_compiler_dock.rb:46:in `sh'
/tmpfs/src/github/protobuf/ruby/Rakefile:116:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-13.0.6/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => gem:native => gem:java
(See full trace by running task with --trace)
```

This seems unrelated to this PR? I'll kick it one more time in case the failure was spurious.

**Comment by acozzette** [comment]
@haberman I believe that last failure is supposed to be fixed by @Maaarcocr's change to rake-compiler-dock, but it sounds like we need to wait for the fix to be released.

**Comment by Maaarcocr** [comment]
> @haberman I believe that last failure is supposed to be fixed by @Maaarcocr's change to rake-compiler-dock, but it sounds like we need to wait for the fix to be released.

 the changes just got released. hopefully this will fix all CIs.

**Comment by acozzette** [comment]
The codespell error is unrelated and I went ahead and fixed it separately.

**Comment by perezd** [comment]
This change is creating fake arm64 ruby gems that are not actually arm64 (reporting as x86_64 Mach-O headers). What should we do about this given our pending release for 21.0?

**Comment by haberman** [comment]
Reference: https://github.com/protocolbuffers/protobuf/issues/9804

Have arm64 binaries on macOS previously worked, and this PR broke them?

**Comment by johnnyshields** [comment]
I have the correct way of doing the `mingw-ucrt` here: https://github.com/protocolbuffers/protobuf/pull/9963. Please kindly merge this so we can have Ruby 3.1 on Windows, thanks!

**Pull request merged**
